src/NUnitFramework/tests/Constraints/DictionaryContainsKeyConstraintTests.cs

```diff
@@ -84,35 +85,521 @@ public void WorksWithNonGenericDictionary()
+
+            Assert.That(lookup, Does.ContainKey("catherine"));
+        }
+
+        [Test]
+        public void SucceedsWhenKeyIsNotPresentUsingContainKeyUsingLookupDefaultComparer()
+        {
+            var list = new List<string> { "ALICE", "BOB", "CATHERINE" };
+            ILookup<string, string> lookup = list.ToLookup(x => x);
+
+            Assert.That(lookup, !Does.ContainKey("alice"));
+        }
+
+        [Test]
+        public void SucceedsWhenKeyIsPresentUsingContainKeyWhenUsingKeyedCollectionCustomComparer()
+        {
+            var list = new TestKeyedCollection(StringComparer.OrdinalIgnoreCase) { "ALICE", "BOB", "CALUM" };
+
+            Assert.That(list, Does.ContainKey("calum"));
+        }
+
+        [Test]
+        public void SucceedsWhenKeyIsNotPresentUsingContainKeyUsingKeyedCollectionDefaultComparer()
+        {
+            var list = new TestKeyedCollection {"ALICE", "BOB", "CALUM"};
+            
+            Assert.That(list, !Does.ContainKey("alice"));
+        }
+
+        [Test]
+        public void SucceedsWhenKeyIsPresentUsingContainKeyUsingHastableCustomComparer()
+        {
+            var table = new Hashtable(StringComparer.OrdinalIgnoreCase) { { "ALICE", "BOB" }, { "CALUM", "DENNIS" } };
+
+            Assert.That(table, Does.ContainKey("alice"));
+        }
+
+        [Test]
+        public void SucceedsWhenKeyIsPresentUsingContainKeyUsingHastableDefaultComparer()
+        {
+            var table = new Hashtable {{ "ALICE", "BOB"}, {"CALUM", "DENNIS"} };
+
+            Assert.That(table, !Does.ContainKey("calum"));
+        }
+
+        [Test]
+        public void ShouldCallContainsKeysMethodWithTKeyParameterOnNewMethod()
+        {
+            var dictionary = new TestDictionaryGeneric<string, string> { { "ALICE", "BOB" }, { "CALUM", "DENNIS" } };
+
+            Assert.That(dictionary, Does.ContainKey("BOB"));
+        }
+
+        [Test]
+        public void ShouldCallContainsKeysMethodOnDictionary()
+        {
+            var dictionary = new TestDictionary(20);
+
+            Assert.That(dictionary, Does.ContainKey(20));
+            Assert.That(dictionary, !Does.ContainKey(10));
+        }
+
+        [Test]
+        public void ShouldCallContainsKeysMethodOnPlainDictionary()
+        {
+            var dictionary = new TestPlainDictionary(99);
+
+            Assert.Catch<ArgumentException>(() => Assert.That(dictionary, Does.ContainKey(99)));
+        }
+
+        [Test]
+        public void ShouldCallContainsKeysMethodOnObject()
+        {
+            var poco = new TestPlainObjectContainsKey("David");
+
+            Assert.Catch<ArgumentException>(() => Assert.That(poco, Does.ContainKey("David")));
+        }
+
+        [Test]
+        public void ShouldCallContainsMethodOnObject()
+        {
+            var poco = new TestPlainObjectContains("Peter");
+
+            Assert.Catch<ArgumentException>(() => Assert.That(poco, Does.ContainKey("Peter")));
```

Comment:
We should have a test on a `TestPlainObjectContains<TKey>` class where this succeeds, too.

Reply:
> 

This is correct functionality... although we could extend the test scenario (read first and decide).

Basically in CollectionConstraint.ApplyTo<>(...) we call a method ConstraintUtils.RequiresActual<IEnmerable>(...) This throws the ArgumentException as our TestNonGenericObjectContrainsKey (new name not PlainObject as before) doesn't implement this interface. This is meant to be a "Collection" of some type, so it seems perfectly valid to throw.

We could make this TestNonGenericObjectContrainsKey impl IEnumerable and then obviously we should be good? Could create a new test with new object inherits this class and which has IEnumerable. I think both seem valid?